src/client/Pipeline.cpp

```diff
@@ -241,18 +271,41 @@ void PipelineImpl::buildForAppendOrRecovery(bool recovery) {
              */
             if (stage != PIPELINE_SETUP_CREATE && stage != PIPELINE_CLOSE
                     && static_cast<int>(nodes.size()) < replication && canAddDatanode) {
-                if (!addDatanodeToPipeline(excludedNodes)) {
+                // Single data node case
+                bool added = false;
+                if (nodes.empty() && useRemoved) {
+                    if (storageID.length()) {
+                        LOG(INFO, "Pipeline: Adding back only datanode %s", removed.formatAddress().c_str());
+                        nodes.push_back(removed);
```

Comment:
yeah, maybe I'm just missing something.
a block on hdfs is associated with a sequential pipeline of replica nodes.
when the client calls `PipelineImp.send` and hdfs fails to sequentially propagate a packet to all the nodes,
it gets `ack.IsInvalid == true` and calculates the first failed node in the pipeline and assigns `errorIndex`.
then there is a `buildForAppendOrRecovery` call when we traverse the pipeline from the end and try to find additional/replacement nodes. as a result, we have updated `nodes` and `storageIds` and submit the changes using `filesystem->updatePipeline`.
my point here is that we call `filesystem->updatePipeline` twice in the body of this function. I'm curious whether it is an issue.